sdk/program/src/sysvar/slot_hashes.rs

```diff
@@ -62,13 +71,110 @@ impl Sysvar for SlotHashes {
     }
 }
 
+#[derive(Copy, Clone, Pod, Zeroable)]
+#[repr(C)]
+struct PodSlotHash {
+    slot: Slot,
+    hash: Hash,
+}
+
+/// API for querying the `SlotHashes` sysvar.
+pub struct SlotHashesSysvar;
+
+impl SlotHashesSysvar {
+    /// Get a value from the sysvar entries by its key.
+    /// Returns `None` if the key is not found.
+    pub fn get(slot: &Slot) -> Result<Option<Hash>, ProgramError> {
+        let data_len = SlotHashes::size_of();
+        let mut data = vec![0u8; data_len];
+
+        // Ensure the created buffer is aligned to 8.
+        // This should never throw unless `SlotHashes::size_of` changes.
+        if data.as_ptr().align_offset(8) != 0 {
+            return Err(ProgramError::InvalidAccountData);
+        }
```

Comment:
Sorry, this check still worries me, because we're relying on this particular allocation strategy for `RawVec`: https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#130

We shouldn't depend on the implementation details of the inner `Vec`, otherwise we might get screwed if it changes, similar to the `StableVec` that we've had to do.

How about using some other type in between which is guaranteed to align to 8 and then casting a slice of that to a byte slice? Here are some examples:

```
    #[repr(C, align(64))]
    #[derive(Clone, Copy, Default, Pod, Zeroable)]
    struct Big{
        a: [u8; 32],
        b: [u8; 32],
    }

    #[test]
    fn alignment() {
// the base case, my machine aligns this to 16 for some reason
        let first = vec![0u8; 8];
        assert_eq!(first.as_ptr().align_offset(16), 0);

// this is a bit more exotic, if needed for bigger types
        let mut first = vec![Big::default(); 10];
        assert_eq!(first.as_ptr().align_offset(64), 0);
        assert_eq!(bytemuck::cast_slice_mut::<Big, u8>(&mut first).as_ptr().align_offset(64), 0);
       
// this one might be simplest
        let mut u64s = vec![0u64; 10];
        assert_eq!(bytemuck::cast_slice_mut::<u64, u8>(&mut u64s).as_ptr().align_offset(8), 0);
    }
```

What do you think? I lean towards the last one, using a vec of `u64`s as the base to make sure it aligns as expected

Reply:
Wouldn't the `PodSlotHash` struct I've defined here align to 8 as well? Then I can just shave off the first 8 bytes for the vector length in the call to `get_sysvar`, and the slice gets basically mutated in-place!

Reply:
yeah that works too!

Reply:
Something like:

```rust
let sysvar_data_len = SlotHashes::size_of() - 8;
let elements = sysvar_data_len / std::mem::size_of::<PodSlotHash>();
let mut layout = vec![PodSlotHash::default(); elements];
let mut data = bytemuck::cast_slice_mut::<PodSlotHash, u8>(&mut layout);
get_sysvar(&mut data, /* offset */ 8, /* length */ sysvar_data_len)?;
```

Reply:
There might even be a more elegant way using bytemuck to go from `&[PodSlotHash]` -> `&[u8]` -> `&[PodSlotHash]`.

Reply:
Yeah that's true, you can even keep it simpler using `vec![PodSlotHash::default(); slot_hashes::MAX_ENTRIES]`, cast it just for the syscall, then go back to using the original vec!

Reply:
Very slick. I pushed that up.